### PR TITLE
GRF-448 : Update end to end tests to cover the with enriched values use case

### PR DESCRIPTION
**Pull request by jeremyBeaucousinAkeneo**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ApiCategoryTestCase.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+namespace Akeneo\Test\Category\EndToEnd\ExternalApi;
+
+use Akeneo\Category\Domain\ValueObject\Attribute\Value\AbstractValue;
+use Akeneo\Tool\Bundle\ApiBundle\tests\integration\ApiTestCase;
+
+abstract class ApiCategoryTestCase extends ApiTestCase
+{
+    protected function activateEnrichedFeatureFlag(): void
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ApiCategoryTestCase.php:17]:
```suggestion
    protected function enableEnrichedCategoryFeature(): void
```

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php [hunk 2]
```diff
@@ -280,6 +279,37 @@ public function testListCategoriesWithPosition(): void
...
+
+        $categories = $this->getStandardizedCategorieswithAttributesValues();
+        $expected = <<<JSON
+{
+    "_links": {
+        "self": {
+            "href": "http://localhost/api/rest/v1/categories?page=1&limit=10&with_count=false&with_enriched_attributes=0&sqlWhere=category.code%20IN%20%28:category_codes%29&sqlLimitOffset=LIMIT%2010&params%5Bcategory_codes%5D%5B0%5D=master&params%5Bwith_enriched_attributes%5D=1&types%5Bcategory_codes%5D=102&types%5Bwith_enriched_attributes%5D=5"
+        },
+        "first": {
+            "href": "http://localhost/api/rest/v1/categories?page=1&limit=10&with_count=false&with_enriched_attributes=0&sqlWhere=category.code%20IN%20%28:category_codes%29&sqlLimitOffset=LIMIT%2010&params%5Bcategory_codes%5D%5B0%5D=master&params%5Bwith_enriched_attributes%5D=1&types%5Bcategory_codes%5D=102&types%5Bwith_enriched_attributes%5D=5"
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:298]: the generated URL looks weird and seems to expose the "public" parameters used for the URL (which are URL encoded) along with the "private" ones used for building the SQL query.
Another point, the first parameter "with_enriched_attributes" equals 0, is it normal?

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php [hunk 3]
```diff
@@ -280,6 +279,37 @@ public function testListCategoriesWithPosition(): void
         $this->assertSameResponse($expected, $client->getResponse());
     }
 
+    public function testListCategoriesWithValues(): void
+    {
+        $this->activateEnrichedFeatureFlag();
+        $this->updateCategoryWithValues('master');
+        $client = $this->createAuthenticatedClient();
+
+        $client->request('GET', 'api/rest/v1/categories?search=["master"]&with_enriched_attributes=true');
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:288]: the search filter does not seem to follow the API reference: https://api.akeneo.com/documentation/filter.html#by-category-codes

The URL should follow this format: `/api/rest/v1/categories?search={"code":[{"operator":"IN","value":["category_code1","category_code2"]}]}`

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php [hunk 4]
```diff
@@ -501,6 +531,66 @@ public function getStandardizedCategorieswithPositionInformation(): array
...
+{
+    "_links": {
+        "self": {
+            "href": "http://localhost/api/rest/v1/categories/master"
+        }
+    },
+    "code": "master",
+    "parent": null,
+    "updated" : "2016-06-14T13:12:50+02:00",
+    "position": null,
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:549]: position is optional

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php [hunk 5]
```diff
@@ -501,6 +531,66 @@ public function getStandardizedCategorieswithPositionInformation(): array
...
+    "labels": null,
+    "template": null,
+    "values": {$this->getStandardizedAttributesValues()}
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:552]: templates and values are optional. They are returned only when the parameter `with_enriched_values` is present

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php [hunk 6]
```diff
@@ -501,6 +531,66 @@ public function getStandardizedCategorieswithPositionInformation(): array
...
+    "_links": {
+        "self": {
+            "href": "http://localhost/api/rest/v1/categories/master"
+        }
+    },
+    "code": "master",
+    "parent": null,
+    "updated" : "2016-06-14T13:12:50+02:00",
+    "position": null,
+    "labels": null,
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:550]: Labels should be an empty list when there is no translated labels for the category

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php [hunk 7]
```diff
@@ -501,6 +531,66 @@ public function getStandardizedCategorieswithPositionInformation(): array
         return $categories;
     }
 
+    /**
+     * @return array<string, mixed>
+     */
+    public function getStandardizedCategorieswithAttributesValues(): array
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:537]: the name let us think it returns a list of categories. We return only the response for the master category

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ApiCategoryTestCase.php [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+use Akeneo\Category\Domain\ValueObject\Attribute\Value\AbstractValue;
+use Akeneo\Tool\Bundle\ApiBundle\tests\integration\ApiTestCase;
+
+abstract class ApiCategoryTestCase extends ApiTestCase
+{
+    protected function enableEnrichedCategoryFeature(): void
+    {
+        $featureFlags = $this->get('feature_flags');
+        $featureFlags->enable('enriched_category');
```
Comment (BryanS29-akeneo) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ApiCategoryTestCase.php:20]: you can directly do this:
```
$this->get('feature_flags')->enable('enriched_category');
```

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php [hunk 9]
```diff
@@ -280,6 +279,37 @@ public function testListCategoriesWithPosition(): void
         $this->assertSameResponse($expected, $client->getResponse());
     }
 
+    public function testListCategoriesWithValues(): void
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:282]: What do you think about adding a [group](https://phpunit.readthedocs.io/en/9.5/annotations.html#group) annotation in order to easily identify the tests related to the Enriched Category feature?

Comment (jeremyBeaucousinAkeneo) [reply src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:282]: Good idea, I was searching for a way to do that, a group is a good idea.

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php (hunk 2 above)
Comment (jeremyBeaucousinAkeneo) [reply src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php]: There were a mix up with sql Parameters and query parameters. I fixed it in isolated SQL parameters.

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php (hunk 3 above)
Comment (jeremyBeaucousinAkeneo) [reply src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:288]: Will be fixed by [GRF-376](https://akeneo.atlassian.net/browse/GRF-376).
This url was constructed to match the existing implementation.

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php (hunk 5 above)
Comment (jeremyBeaucousinAkeneo) [reply src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php]: Json is paramatized with a boolean. 
PHP`public function getStandardizedCategorieswithAttributesValues(bool $withEnrichedValues): array`
We can discuss that tomorrow if you've got a better idea

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php (hunk 7 above)
Comment (jeremyBeaucousinAkeneo) [reply src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php]: I use the \Akeneo\Test\Category\EndToEnd\ExternalApi\ListCategoryEndToEnd::getStandardizedCategorieswithPositionInformation with parameters to generate json

File: src/Akeneo/Category/back/ServiceApi/ExternalApiCategory.php [hunk 10]
```diff
@@ -107,16 +107,21 @@ public static function fromDomainModel(CategoryDomain $category): self
...
     {
-        return [
+        $normalizedCategory = [
             'code' => $this->getCode(),
             'parent' => $this->getParentCode(),
             'updated' => $this->getUpdated(),
-            'position' => $this->getPosition(),
             'labels' => $this->getLabels(),
             'template' => $this->getTemplateCode(),
             'values' => $this->getValues(),
```
Comment (BryanS29-akeneo) [inline src/Akeneo/Category/back/ServiceApi/ExternalApiCategory.php:118]: I think you can do the same like you do for "with_position" with "values" which is optional (`with_enriched_attributes` parameter)

File: src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php [hunk 11]
```diff
@@ -289,229 +311,173 @@ protected function getConfiguration(): Configuration
     }
 
     /**
-     * @return array<string,mixed>
+     * @return array<string, mixed>
      */
-    public function getStandardizedCategories(): array
+    public function getStandardizedCategorieswithPositionInformation(bool $withPosition, bool $withEnrichedValues): array
```
Comment (BryanS29-akeneo) [inline src/Akeneo/Category/back/tests/EndToEnd/ExternalApi/ListCategoryEndToEnd.php:316]: you can rename the name of the function, it doesn't concern position only

**Review by BryanS29-akeneo — APPROVED** [review]

**Review by etiennekerbastard — APPROVED** [review]

**Pull request merged**